src/briefcase/platforms/android/apk.py

```diff
@@ -1 +1,235 @@
+in the first column of output.
+
+$ {adb} devices -l
+"""
+
+
+class ApkRunCommand(ApkMixin, RunCommand):
+    description = "Run an Android APK."
+
+    def verify_tools(self):
+        super().verify_tools()
+        if not (self.android_sdk_path / "emulator").exists():
+            print("Ensuring we have the Android emulator and system image...")
+            # TODO: Error handling.
+            self.subprocess.run([
+                self.android_sdk_path / "tools" / "bin" / "sdkmanager",
+                "platforms;android-28",
+                "system-images;android-28;default;x86",
+                "emulator",
+                "platform-tools",
+            ])
+
+    def add_options(self, parser):
+        super().add_options(parser)
+        parser.add_argument(
+            '-d',
+            '--device',
+            dest='device',
+            help='The device to target, formatted for `adb`',
+            required=False,
+        )
+
+    def run_app(self, app: BaseConfig, device=None, **kwargs):
+        """
+        Start the application.
+
+        :param app: The config object for the app
+        :param device: The device to target. If ``None``, the user will
+            be asked to re-run the command selecting a specific device.
+        :param base_path: The path to the project directory.
+        """
+        if device is None:
+            raise BriefcaseCommandError("""\
+Please specify a specific device on which to run the app by passing
+`-d device_name`.""".lstrip().format(
+                adb=self.android_sdk_path / "platform-tools" / "adb",
+                emulator=self.android_sdk_path / "emulator" / "emulator",
+                tools_bin=self.android_sdk_path / "tools" / "bin",
+            ))
+
+        # Install the latest APK file onto the device.
+        try:
+            # TODO: Decide how to handle general BriefcaseCommandError.
+            install_apk(self.android_sdk_path, device, self.binary_path(app))
+        except DeviceNotFound:
+            print("Device {device} not found.".format(device=device))
+            print("")
+            raise BriefcaseCommandError(NO_OR_WRONG_DEVICE_MESSAGE.format(
+                adb=self.android_sdk_path / "platform-tools" / "adb",
+                emulator=self.android_sdk_path / "emulator" / "emulator",
+                tools_bin=self.android_sdk_path / "tools" / "bin",
+            ))
+
+        # Compute Android package name based on beeware `bundle` and `app_name`
+        # app properties, similar to iOS.
+        package = "{app.bundle}.{app.app_name}".format(app=app)
+
+        # We force-stop the app to ensure the activity launches freshly.
+        force_stop_app(self.android_sdk_path, device, package)
+
+        # To start the app, we launch `org.beeware.android.MainActivity`.
+        start_app(
+            self.android_sdk_path, device, package,
+            "org.beeware.android.MainActivity"
+        )
```

Comment:
Do these commands have failure modes?

Reply:
`start_app` does, yeah. I can try to exercise those failure modes and write tests that cover them.

Reply:
Ah - I see the underlying integration is raising the BriefcaseError.